### DISCUSSION TITLE
How sandboxed are applications on multi-user systems?

### DISCUSSION THREAD

**Discussion by milesromeno** [Homebrew/discussions#5032] [Everyday usage]
Apologies if this is not the right forum to ask, I am not sure if my question is primarily related to homebrew or to MacOS in general.

### Description of issue

The recommended approach is to have a single brew installation at `opt/homebrew` on Apple Sillicon, and the default cask `--appdir` is `/Applications` rather than `~/Applications`. 

A result of this I have noticed is that if I login to user `A` on my Mac and run `brew install google-chrome`, then logout and login to user `B`, Google Chrome will still show up in Spotlight search and can be ran as if `B` had already installed it. However, any information that I had setup while logged in as `A`, like my Gmail account sign-in and search history, does not transfer in as `B`. It's as if the application is being opened for the first time, looking effectively sandboxed for each user. I see this with many apps: I can install `spotify` or `discord` one time and have it accessible as an application for all users on my macbook, but will still have to manually re-log into spotify/discord accounts for each user on first run. The application executable is shared, but the application's save files or caches are apparently not. This is good, I want each user to have total isolation and privacy from other users on the system.

I have a few questions (though answering one probably answers all of them):

* What is technically happening? Does brew install the app executable at one global location, but keep caches/configs/saves at directories local to each user? Or is it the particular applications in question that are choosing to do this?
* Is this behavior specific to installing applications with `brew`, or would I still see the same sandboxing effect if I manually installed the apps via Apple App Store or using native installers from the app websites?
* Does it work this way for all applications? Can I confidently trust that all applications installed using brew will be sandboxed for each user to make sure no personal details or application login sessions are leaked?
* To what extent is the isolation? Do the application instances on different users know about each other? As in, is Discord capable of correlating a discord account signed in while logged in as `A` with a discord account signed in while logged in as `B`, simply by virtue of the fact that the the application was installed globally using brew?

**Comment by SMillerDev** [comment]
> What is technically happening? Does brew install the app executable at one global location, but keep caches/configs/saves at directories local to each user? Or is it the particular applications in question that are choosing to do this?

It's up to an application to decide where they store their data.

> Is this behavior specific to installing applications with brew, or would I still see the same sandboxing effect if I manually installed the apps via Apple App Store or using native installers from the app websites?

Yes

> Does it work this way for all applications? Can I confidently trust that all applications installed using brew will be sandboxed for each user to make sure no personal details or application login sessions are leaked?

No, especially formula will probably write data to standard UNIX locations under the brew prefix rather than confine it to the user.

> To what extent is the isolation? Do the application instances on different users know about each other? As in, is Discord capable of correlating a discord account signed in while logged in as A with a discord account signed in while logged in as B, simply by virtue of the fact that the the application was installed globally using brew?

Depends entirely on the app.

**Reply by milesromeno** [reply]
When you answered the second question with
> Yes

was that saying yes to the first case of the *or* statement (behavior specific to brew), or the second (same behavior if installed using proprietary installers)? I assume the second from your other answers, just want to confirm.

**Reply by SMillerDev** [reply]
It's the same whichever way you install it.

**Reply by milesromeno** [reply]
So just to confirm, when you said
> No, especially formula will probably write data to standard UNIX locations under the brew prefix rather than confine it to the user.

Does this mean that it is likely for some applications that if I open it as user `A` and sign-in to `someuser@email.com`, user `B` may automatically be signed-in to `someuser@email.com` as well when they open the application for the first time? That seems very worrisome for security.

**Reply by SMillerDev** [reply]
Generally formula there isn't much of a login, but it could happen. It could also be that it writes to `~/.local/share` or something. There is simply no telling since brew has no control over this.
